src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.TrailingHeaders.cs

```diff
@@ -0,0 +1,225 @@
+                            "Transfer-Encoding: chunked\r\n" +
+                            "Trailer: MyCoolTrailerHeader\r\n" +
+                            "\r\n" +
+                            "4\r\n" +
+                            "data\r\n" +
+                            "0\r\n" +
+                            "MyCoolTrailerHeader: amazingtrailer\r\n" +
+                            "Hello: World\r\n" +
+                            "\r\n"));
+
+                    using (HttpResponseMessage response = await getResponseTask)
+                    {
+                        Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+                        Assert.Contains("chunked", response.Headers.GetValues("Transfer-Encoding"));
+                        Assert.Contains("MyCoolTrailerHeader", response.Headers.GetValues("Trailer"));
+
+                        // Pending read on the response content.
+                        Assert.Empty(response.TrailingHeaders);
+
+                        Stream stream = await response.Content.ReadAsStreamAsync();
+                        Byte[] data = new Byte[100];
+                        int start = 0;
+                        await stream.ReadAsync(data, start, 4);
+                        start = 4;
+
+                        // Intermedia test - haven't reached stream EOF yet.
+                        Assert.Empty(response.TrailingHeaders);
+                        Assert.Contains("data", System.Text.Encoding.Default.GetString(data));
+
+                        bool streamEOFReached = false;
+                        while (!streamEOFReached)
+                        {
+                            if (stream.Read(data, start, 10) == 0)
+                            {
+                                streamEOFReached = true;
+                            }
+                            start += 10;
+                        }
+
+                        Assert.Contains("amazingtrailer", response.TrailingHeaders.GetValues("MyCoolTrailerHeader"));
+                        Assert.Contains("World", response.TrailingHeaders.GetValues("Hello"));
+                    }
+                }
+            });
+        }
+
+        [Fact]
+        public async Task GetAsync_ForbiddenTrailingHeaders_Throws()
+        {
+            await LoopbackServer.CreateClientAndServerAsync(async url =>
+            {
+                using (HttpClientHandler handler = CreateHttpClientHandler())
+                using (var client = new HttpClient(handler))
+                {
+                    await Assert.ThrowsAsync<HttpRequestException>(() => client.GetAsync(url));
+                }
+            }, server => server.AcceptConnectionSendCustomResponseAndCloseAsync(
+                "HTTP/1.1 200 OK\r\n" +
+                "Connection: close\r\n" +
+                "Transfer-Encoding: chunked\r\n" +
+                "Trailer: Set-Cookie, MyCoolTrailerHeader, Hello, Content-Type\r\n" +
```

Comment:
There was a much longer list of forbidden trailers.  I suggest making this test a theory, with an InlineData for each forbidden trailer.  It would also be good to validate that other known headers that aren't forbidden are properly accepted.

Reply:
test is updated to be theory.